contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,300 @@
+async function processEvent(event, { config, cache }) {
+    const counterValue = (await cache.get('greeting_counter', 0))
+    cache.set('greeting_counter', counterValue + 1)
+    if (!event.properties) event.properties = {}
+    event.properties['greeting'] = config.greeting
+    event.properties['greeting_counter'] = counterValue
+    event.properties['random_number'] = await getRandomNumber()
+    return event
+}
+
+// The plugin itself
+module.exports = {
+    setupPlugin,
+    processEvent
+}
+```
+
+And here’s a basic `plugin.json` that goes along with it:
+
+```js
+{
+    "name": "Hello World",
+    "url": "https://github.com/PostHog/posthog-hello-world-plugin",
+    "description": "Greet the world with every PostHog event. An example plugin.",
+    "main": "index.js",
+    "config": [
+        {
+            "key": "greeting",
+            "name": "What greeting would you like to use?",
+            "type": "string",
+            "default": "Hello world!",
+            "required": false
+        }
+    ]
+}
+```
+
+### App functions and processes
+
+As you can see by reading the above examples, the index file exports functions that take objects PostHog provides (like configuration or event data). Specifically, our PostHog exports predefined functions that run on conditions like:
+
+- event ingestion (`processEvent`)
+    - useful for modifying incoming data, such as adding geographic data properties or dropping events matching a property filter.
+- every minute, hour, day (`runEveryMinute`, `runEveryHour`, `runEveryDay`)
+    - useful for triggering events based on updates to external services such as Stripe, HubSpot, or GitHub.
+- once processing is complete (`exportEvents`, `onEvent`)
+    - useful for exporting data to external services such as BigQuery, Snowflake, or Databricks.
+
+When users upload apps to PostHog, these we export these functions from the files. From there they run as callable functions on VMs as part of our ingestion pipeline and also have access to extensions like storage, caching, and logging. Overall, it looks like this:
+
+![VM](../images/blog/how-we-built-an-app-server/vm.png)
+
+Structuring apps this way provides flexibility to users while maintaining scalability and security. Added features like scheduling and exporting added new functionality users wanted. Caching and improving the config object (enabling statelessness) helped apps scale effectively. Running apps in VMs helped ensure they were secure and isolated. The next step was to put these apps into production.
+
+## Serving apps and managing data at scale
+
+An app is no good if it doesn't integrate into PostHog. Apps needed to integrate with our ingestion pipeline. They needed to handle an inflow of data, scale, manage tasks, and update themselves. There are many parts needed to make this happen.
+
+Below is a basic diagram of our ingestion pipeline. Most of what we care about is the plugin server and the Kafka data flow leading into it.
```

Comment:
First time "plugin server" is mentioned". What's the difference between plugins and apps?